Fix broken voice mixing

derselbst opened the pull request:
When rendering a voice, there are 3 cases to consider: silence, playing and finished. When optimizing away the `memset`, I incorrectly assumed that a voice cannot switch between playing and silence (like crazy) while rendering `FLUID_MIXER_MAX_BUFFERS_DEFAULT` . Apparently this does not hold true, esp. for percussion instrument while rendering at 96kHz. ~Thus, render one single block and immediately mix it into the destination buffer(s). (Rather than rendering all the `FLUID_MIXER_MAX_BUFFERS_DEFAULT` blocks and then do the mixing once).~

Fixes #580.

derselbst commented:
@jjceresa Pls. confirm the fix.

jjceresa commented:
This doesn't fix. Now clipping occurs at any sample rate. For example when playing `preset 5 (1 based) electric piano 1`, playing C4,D4,C4,D4,.... very quickly from C to D, the two notes are mixed and during the switch from one note to the other we hear clippings. Using `Preset 17 Electric organ 1` or `preset 18 electric organ 2` also shows  the issue.

jjceresa commented:
I think that `fluid_rvoice_buffers_mix()` should still receive `s` param which is the number of sample returned by `fluid_rvoice_write()`. Indeed, It is possible that `s < FLUID_BUFSIZE` and in this case `fluid_rvoice_buffers_mix()` should mix only `s` samples.
Actually `fluid_rvoice_buffers_mix()` always mixes `FLUID_BUFSIZE` samples (i.e one complete block). That means that `(FLUID_BUFSIZE - s)  invalid samples` from the `local buffer` will be mixed to the destination buffer. It seems that this local buffer  was not filled by 0.0, so I feel that this new parameter is still necessary.

jjceresa commented:
>I incorrectly assumed that a voice cannot switch between playing and silence (like crazy) while rendering FLUID_MIXER_MAX_BUFFERS_DEFAULT . Apparently this does not hold true...

Right, when a voice is started, amplitude envelope `DAHDSR  could be in `delay` stage at the beginning. During this time `fluid_rvoice_write()` return -1 to indicate that the voice is quiet.

rendering all the FLUID_MIXER_MAX_BUFFERS_DEFAULT blocks and then do the mixing once shouldn't be more efficient  ? even if  we keep `memset` to clear the local buffer ?.

jjceresa commented:
`>I incorrectly assumed that a voice cannot switch between playing and silence (like crazy) while rendering FLUID_MIXER_MAX_BUFFERS_DEFAULT . Apparently this does not hold true...

Right, when a voice is started, amplitude envelope `DAHDSR` could be in `delay`stage at the beginning. During this time `fluid_rvoice_write()` return -1 to indicate that the voice is quiet.

Rendering all the FLUID_MIXER_MAX_BUFFERS_DEFAULT blocks and then do the mixing once shouldn't be more efficient ? even if we keep `memset` to clear the local buffer ?.

jjceresa commented:
>I incorrectly assumed that a voice cannot switch between playing and silence (like crazy) while rendering FLUID_MIXER_MAX_BUFFERS_DEFAULT . Apparently this does not hold true...

Right, when a voice is started, amplitude envelope `DAHDSR` could be in `delay` stage at the beginning. During this delay `fluid_rvoice_write()` return -1 to indicate that the voice is quiet.

Rendering all the FLUID_MIXER_MAX_BUFFERS_DEFAULT blocks and then do the mixing once shouldn't be more efficient ? even if we keep `memset` to clear the local buffer ?.

derselbst commented:
> This doesn't fix. Now clipping occurs at any sample rate.

You are right. Believe me or not, when I tested it with jack and sdl2, it sounded absolutely fine. After listening to alsa, I heard what you mean.

At least now I know why there are so many differences between the driver: the `blockcount` varies dramatically. Alsa for example (at least on my system) always renders with `blockcount==1`. Jack and sdl2 however render many blocks.

> Rendering all the FLUID_MIXER_MAX_BUFFERS_DEFAULT blocks and then do the mixing once shouldn't be more efficient ?

Yes, that's how I implemented it now.

> even if we keep memset to clear the local buffer ?.

No. When I refactored `rvoice_mixer`, I also profiled it and `memset` was one of the most expensive things.

So, pls. have another try.

jjceresa commented:
>So, pls. have another try.

All is fine.
In last commit, I did simple parameter change, arithmetic simplification, and adding comments.

>Believe me or not, when I tested it with jack and sdl2, it sounded absolutely fine. After listening to alsa, I heard what you mean.

I believe you.

>At least now I know why there are so many differences between the driver: the blockcount varies dramatically.

Yes this blockcount isn't predictive and make debugging very difficult. Using 96kHz I had to change audio.period-size to get the driver working correctly. Probably most driver need to do sample rate conversion too.

derselbst commented:
Thx for testing.

> In last commit, I did simple parameter change, arithmetic simplification, and adding comments.

I prefer `buf[start_block * FLUID_BUFSIZE + dsp_i]`, because it tells the compiler that the start of accessing the `buf` will always be at the `FLUID_BUFSIZE*sizeof(fluid_real_t)` byte border. This matters, because I'm promising that `buf` itself is aligned to `FLUID_DEFAULT_ALIGNMENT`, which avoids the need of a peel loop in case the compiler tries to vectorize it. `buf[start_sample + dsp_i]` , however, can be somewhere in between, requiring a peel loop.

Will add a comment about that.

P.S.: pay attention to the base branch when merging: Base branch of `fix580` is `2.0.x`, not master. Will fix that as well.

jjceresa commented:
>I prefer buf[start_block * FLUID_BUFSIZE + dsp_i], because it tells the compiler that the start of accessing the buf..

Ok, I understand now why you did that. I forgot about alignment requirement.

>P.S.: pay attention to the base branch when merging: Base branch of fix580 is 2.0.x, not master. Will fix that as well.

Ok, sorry for this disorder.

jjceresa commented:
Pls, don't forget to propagate the fix to master. Thanks.
